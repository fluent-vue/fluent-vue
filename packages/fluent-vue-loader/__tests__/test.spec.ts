import type { Stats } from 'webpack'

import compiler from './compiler'

describe('webpack loader', () => {
  test('generates custom block code', async () => {
    // Arrange
    // Act
    const stats = await compiler('fixtures/test.vue')

    // Assert
    const ftlModules = stats.toJson({ source: true }).modules
      ?.filter(module =>
        module.name?.includes('blockType=fluent') === true &&
        module.issuerName?.includes('blockType=fluent') === true)
      .map(module => module.source)

    expect(ftlModules).not.toBeUndefined()
    expect(ftlModules).toHaveLength(1)
    expect(ftlModules).toMatchSnapshot()
  })

  test('includes HMR code only when HMR is enabled', async () => {
    // Arrange
    // Act
    const stats = await compiler('fixtures/test.vue', {}, true)

    // Assert
    const ftlModules = stats.toJson({ source: true }).modules
      ?.filter(module =>
        module.name?.includes('blockType=fluent') === true &&
        module.issuerName?.includes('blockType=fluent') === true)
      .map(module => module.source)

    expect(ftlModules).not.toBeUndefined()
    expect(ftlModules).toHaveLength(1)
    expect(ftlModules).toMatchSnapshot()
  })

  test('errors with no locale attr', async () => {
    // Arrange
    const func = async (): Promise<Stats> => await compiler('fixtures/noLocale.vue')

    // Act
    // Assert
    await expect(func).rejects.toContainEqual(
      expect.objectContaining({
        details: expect.stringContaining('Error: Custom block does not have locale attribute')
      }))
  })
})
