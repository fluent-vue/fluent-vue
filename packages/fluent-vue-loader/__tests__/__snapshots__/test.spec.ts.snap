// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack loader generates custom block code 1`] = `
Array [
  "
import { FluentResource } from '@fluent/bundle'

export default function (Component) {
  const target = Component.options || Component
  target.fluent = target.fluent || {}
  target.fluent['en'] = new FluentResource(\`# Simple things are simple.
hello-user = Hello, {$userName}!

# Complex things are possible.
shared-photos =
  {$userName} {$photoCount ->
     [one] added one photo
    *[other] added {$photoCount} new photos
  } to {$userGender ->
     [male] his stream
     [female] her stream
    *[other] their stream
  }.\`)
  
}
",
]
`;

exports[`webpack loader includes HMR code only when HMR is enabled 1`] = `
Array [
  "
import { FluentResource } from '@fluent/bundle'

export default function (Component) {
  const target = Component.options || Component
  target.fluent = target.fluent || {}
  target.fluent['en'] = new FluentResource(\`# Simple things are simple.
hello-user = Hello, {$userName}!

# Complex things are possible.
shared-photos =
  {$userName} {$photoCount ->
     [one] added one photo
    *[other] added {$photoCount} new photos
  } to {$userGender ->
     [male] his stream
     [female] her stream
    *[other] their stream
  }.\`)
  if (module.hot) {
    delete target._fluent
    if (typeof __VUE_HMR_RUNTIME__ !== 'undefined') {
      // Vue 3
      const id = target.__hmrId
      const api = __VUE_HMR_RUNTIME__
      api.reload(id, target)
    } else {
      // Vue 2
      // There is no proper api to access HMR for component from custom block
      // so use this magic
      delete target._Ctor
    }
  }
}
",
]
`;
